docs/reference-manual/native-image/DebugInfo.md

```diff
@@ -772,4 +789,158 @@ For example, calling the method below prints high-level information about the Na
+            |
+            ---_start
+               __libc_start_main_alias_2 (inlined)
+               |          
+               |--65.21%--__libc_start_call_main
+               |          com.oracle.svm.core.code.IsolateEnterStub::JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b (inlined)
+               |          com.oracle.svm.core.JavaMainWrapper::run (inlined)
+               |          |          
+               |          |--55.84%--com.oracle.svm.core.JavaMainWrapper::runCore (inlined)
+               |          |          com.oracle.svm.core.JavaMainWrapper::runCore0 (inlined)
+               |          |          |          
+               |          |          |--55.25%--DirTest::main (inlined)
+               |          |          |          |          
+               |          |          |           --54.91%--DirTest::listAll (inlined)
+               . . .
+```              
+
+Unfortunately, other operations require Java methods to be identified
+by an ELF (local) function symbol table entry locating the start of
+the compiled method code.
+In particular, assembly code dumps provided by both tools identify
+branch and call targets using an offset from the nearest symbol.
+Omitting Java method symbols means that offsets are generally
+displayed relative to some unrelated global symbol, usually the entry
+point for a method exported for invocation by C code.
+
+As an illustration of the problem, the following excerpted output from
+`perf annotate` displays the first few annotated instructions of the
+compiled code for method `java.lang.String::String()`.
+
+```
+    . . .
+         : 501    java.lang.String::String():
+         : 521    public String(byte[] bytes, int offset, int length, Charset charset) {
+    0.00 :   519d50: sub    $0x68,%rsp
+    0.00 :   519d54: mov    %rdi,0x38(%rsp)
+    0.00 :   519d59: mov    %rsi,0x30(%rsp)
+    0.00 :   519d5e: mov    %edx,0x64(%rsp)
+    0.00 :   519d62: mov    %ecx,0x60(%rsp)
+    0.00 :   519d66: mov    %r8,0x28(%rsp)
+    0.00 :   519d6b: cmp    0x8(%r15),%rsp
+    0.00 :   519d6f: jbe    51ae1a <graal_vm_locator_symbol+0xe26ba>
+    0.00 :   519d75: nop
+    0.00 :   519d76: nop
+         : 522    Objects.requireNonNull(charset);
+    0.00 :   519d77: nop
+         : 524    java.util.Objects::requireNonNull():
+         : 207    if (obj == null)
+    0.00 :   519d78: nop
+    0.00 :   519d79: nop
+         : 209    return obj;
+    . . .
+```
+
+The leftmost column shows percentages for the amount of time recorded
+at each instruction in samples obtained during the `perf record` run.
+Each instruction is prefaced with it's address in the program's code
+section.
+The disassembly interleaves the source lines from which the code is
+derived, 521-524 for the top level code and 207-209 for the code
+inlined from from `Objects.requireNonNull()`.
+(note, however that the lines for the quoted comments are being computed
```

Comment:
s/lines/assembly lines

Reply:
Actually, the assembly lines are always going to be a bit out wrt to the source but but that is not what I meant. This comment is in the wrong place, which explains why you thought I was saying something different.

It is actually the line *numbers* in the trailing javadoc for the method in the *second* sample that are wrong. You can see that in this sample the first line of the method is correctly identified as being at 521. In the second sample that line is output preceded by the 5 trailing lines of the javadoc. For some reason perf has started numbering the trailing lines from 521 rather than from 516. This looks like a perf issue rather than a DWARF one.

I'll relocate the comment to after the second sample and adjust it to say that the line numbering has been applied incorrectly.